File: chap6/AdamsBashforth.f90
!---------------------------------------------------------
!             此程序用的是AdamsBashforth(多步外插法)求解
!                       y'=-3*y
!                       y(0) = 1
!                       t = [0,2]
!                   解析解：y = e^(-3t)
!---------------------------------------------------------
Module AdamsBashforth
  Implicit none
Contains
  Subroutine solve()
    Implicit none
    Real(kind=8), parameter :: h = 1d-2, T0 = 0.d0, T1 = 2.d0
    Integer, parameter :: n = 3, num = nint( (T1-T0)/h )  !// n=3,(0,1,2,3)表示四阶四步法
    Real(kind=8) :: a(0:n), b(0:n), sum, t, sum1, b1, jL, u(0:n), u1(0:num)
    Integer :: i, j, L
  
    Write(*,*) "----------求解Adams外插公式系数值-----------"
    a(0) = 1.d0
    Do i = 1, n
      sum = 0.d0; t = 2.d0
      Do j = i, 1, -1
        sum = sum + a(j-1) / (t)
        t = t + 1.d0
      End do
      a(i) = 1.d0 - sum
    End do
    Write(*,*) a
  
    Write(*,*) "----------验证Adams外插公式系数值-----------"
    Print*
    !//验证Adams外插公式系数值分解是否正确
    sum1 = 0.d0
    Do i = 0, n
      sum1 = sum1 + a(n-i) / (1+i)
    End do
    If( abs(sum1-1.0) < 1d-6 ) then   !//sum1=1.0,说明Adams外插公式系数分解正确
      write(*,*) "----------Adams外插公式系数分解正确---------"  
    else
      write(*,*) "---------Adams外插公式系数分解不正确--------"
      stop
    End If
  
    Print*, '--------------------------------------------'
    Write(*,*) "---------------求解系数Bkl值----------------"
    Print*, '--------------------------------------------'
    Write(*,*) 
  
    Do L = 0, n
      b1 = 0.0
      Do j = L, n
        call calculate_jL(j,L,jL)  !//求解(s(s-1)...(s-j+1)/j!)
        b1 = b1 + (-1)**L * a(j) * jL
      End do
      b(L) = b1
    End do
    Write(*,*) b*24
  
    !//由梯形算法计算前四个解
    call Trapezoid( n, u1, num, h, t0 )
  
    !//AdamsForth公式
    Do i = n+1, num
      u1(i) = u1(i-1) + h * ( b(0)*(-3.*u1(i-1)) + b(1)*(-3.*u1(i-2)) + b(2)*(-3.*u1(i-3)) + b(3)*(-3.*u1(i-4)) )
    End do
  
    open( 100, file = 'waicha.dat', status = 'unknown' )
    Do i = 0, num
      write( 100,'(4f12.8)' ) i*h, u1(i), exp(-3.d0*h*i), abs((u1(i)-exp(-3.d0*h*i)) / exp(-3.d0*h*i)) * 100
    End do
    close( 100 ) 
  End subroutine solve
  
  Subroutine calculate_jL( j, L, jL )
    Implicit none
    Integer :: i
    Integer, intent(in) :: j, L
    Real(kind=8), intent(inout) :: jL
    Real(kind=8) :: Product1, Product2
 
    If( L == 0 ) then
      jL = 1.d0
      return
    else
      Product1 = 1.d0; Product2 = 1.d0
      Do i = 1, L
        Product1 = Product1 * dble(i)  !//求解分子s(s-1)...(s-j+1)
      End Do
      Do i = j, j-L+1, -1
        Product2 = Product2 * dble(i)  !//求解j!
      End Do
      jL = Product2 / Product1
    End If
  End Subroutine calculate_jL
  
  Subroutine Trapezoid( n, u1, num, h, t0 )
    Implicit none
    Integer, intent(in) :: n, num
    Real(kind=8), intent(inout) :: u1(0:n)
    Real(kind=8), intent(in) :: h, t0
    Real(kind=8) :: y0, t, y
    Integer :: i
    
    y0 = 1.d0  !// 初值
    u1(0) = y0
    Do i = 1, n
      t = t0 + h * ( i - 1 )
      !// Wi+1 = Wi + h * ( f(Ti,Wi) + f( Ti+h,Wi+h*f(Ti,Wi) ) )
      y = y0 + h * ( derfunc( y0 ) + derfunc( y0 + h*derfunc( y0 ) ) ) / 2.d0
      y0 = y
      u1(i) = y
    End do
  End subroutine Trapezoid
  
  Real(kind=8) function derfunc( y )
    Implicit none
    Real(kind=8), intent(in) :: y
    derfunc = -3.d0 * y  !// 已知的微分方程
  End function derfunc
  
End module AdamsBashforth
  
  
Program main
  use AdamsBashforth
  Implicit none
  call solve()
End Program main

File: chap6/AdamsBashforth.py
# -*- coding: utf-8 -*-
"""
Created on Tue May 22 19:16:15 2018

@author: luk
"""

"""
y'=-3*y
y(0) = 1
t = [0,2]
解析解：y = e^(-3t)
"""

import numpy as np
t0 = 0.0; t1 = 2.0  # 求值区间
h = 0.01 # 步长
y0 = 1.0
n = 3  # 已知前四个解 n=3,(0,1,2,3)表示四阶四步法
num = round( (t1-t0)/h )
eps = 1e-6

a = np.zeros( (n+1,), dtype = np.float64 )
b = np.zeros( (n+1,), dtype = np.float64 )
x = np.zeros( (num+1,), dtype = np.float64 )
u1 = np.zeros( (num+1,), dtype = np.float64 )  
u2 = np.zeros( (num+1,), dtype = np.float64 )  

print( "求解Adams外插公式系数值" )
a[0] = 1.0
for i in range(1,n+1):
    s = 0.0; t = 2.0
    for j in range(i,0,-1):
        s = s + a[j-1] / t
        t = t + 1.0
    
    a[i] = 1.0 - s  
print( a ) 
  
print( "验证Adams外插公式系数值" )
sum1 = 0.0
for i in range(n+1):
    sum1 = sum1 + a[n-i] / (1+i)
    
if ( abs(sum1-1.0) < eps ):  # sum1=1.0,说明Adams外插公式系数分解正确
    print( "Adams外插公式系数分解正确" )
else:
    print( "Adams外插公式系数分解不正确" )
    quit()
  
def calculate_jL( j, L ):
    global jL
    
    if ( L == 0 ):
        jL = 1.0
        return
    else:
        product1 = 1.0; product2 = 1.0
        for i in range(1,L+1):
            product1 = product1 * np.float64(i)  # 求解分子s(s-1)...(s-j+1)
            
        for i in range(j,j-L,-1):
            product2 = product2 * np.float64(i)  # 求解j!
            
        jL = product2 / product1

    return
    
print( "求解系数Bkl值" )
for L in range(n+1):
    b1 = 0.0
    for j in range(L,n+1):
        jL = 0.0
        calculate_jL(j,L)  # 求解(s(s-1)...(s-j+1)/j!)
        b1 = b1 + (-1)**L * a[j] * jL
    b[L] = b1
    
print( b*24 )  # 乘以24只是为了显示为整数

# 系数求取完毕，接下来用梯形法求取前四个数
derfunc = lambda y: -3.0 * y  # 由初值问题而定

def Trapezoid( n, h, t0 ):
    global u1, u2, y0, x
    
    u1[0] = y0; u2[0] = y0
    for i in range(1,n+1):
        t = t0 + h* (i-1)
        y = y0 + h * ( derfunc( y0 ) + derfunc( y0 + h*derfunc( y0 ) ) ) / 2.0
        y0 = y
        u1[i] = y
        x[i] = t + h
        u2[i] = np.exp( -3.0*x[i] )

Trapezoid( n, h, t0 )  # 求取前四个值     
# AdamsForth公式求解
for i in range(n+1,num+1):
    x[i] = t0 + i * h
    u2[i] = np.exp( -3.0*x[i] )  # 解析解
    u1[i] = u1[i-1] + h * ( b[0]*(-3.0*u1[i-1]) + b[1]*(-3.0*u1[i-2]) + b[2]*(-3.0*u1[i-3]) + b[3]*(-3.0*u1[i-4]) )  # 数值解
    
    
# 绘图
from matplotlib import pyplot as plt
fig, axes = plt.subplots()
axes.plot(x, u1, 'r')
axes.plot(x, u2, 'b--')

plt.title( "AdamsBashforth Method" )  # 设置图标题
plt.xlabel("x")  # 设置坐标轴名称
plt.ylabel("y")

axes.legend( ["u1","u2"], loc = 9 )  # 设置图例, loc = 9 将legend的位置置于顶部中间

# 调整loc的大小可改变其位置，loc的范围[0-10]
plt.show()

# 绘制误差图
fig, axes = plt.subplots()
axes.plot(x, (u1-u2)/u2*100, 'r-.')   # 外插百分比误差

plt.title( "AdamsBashforth Error" )  # 设置图标题
plt.xlabel("x")  # 设置坐标轴名称
plt.ylabel("y")

axes.legend( ["AdamsBashforth Error"], loc = 2 )  # 设置图例, loc = 9 将legend的位置置于顶部中间
plt.show()

print( "please input Enter and  stop!" )
input()   
    
        
    